plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/MavenUpgrade.java

```diff
@@ -140,24 +163,30 @@ void dumpArtifacts(Path file) throws ProvisioningException, MojoExecutionExcepti
-                builder.append("        <artifactId>").append(a.getArtifactId()).append("</artifactId>").append(System.lineSeparator());
-                if (a.getClassifier() != null && !a.getClassifier().isEmpty()) {
-                    builder.append("        <classifier>").append(a.getClassifier()).append("</classifier>").append(System.lineSeparator());
+        builder.append("  <feature-packs>").append(System.lineSeparator());
+        for (Entry<String, Map<String, Map<String, String>>> fp : perFeaturePack.entrySet()) {
+            builder.append("    <feature-pack producer=\"").append(fp.getKey()).append("\" >").append(System.lineSeparator());
+            builder.append("      <modules>").append(System.lineSeparator());
+            for (Entry<String, Map<String, String>> module : fp.getValue().entrySet()) {
+                builder.append("        <module name=\"").append(module.getKey()).append("\">").append(System.lineSeparator());
+                for (String s : module.getValue().values()) {
+                    Artifact a = AbstractBuildBootableJarMojo.getArtifact(s);
+                    builder.append("          <artifact>").append(System.lineSeparator());
+                    builder.append("            <groupId>").append(a.getGroupId()).append("</groupId>").append(System.lineSeparator());
+                    builder.append("            <artifactId>").append(a.getArtifactId()).append("</artifactId>").append(System.lineSeparator());
+                    if (a.getClassifier() != null && !a.getClassifier().isEmpty()) {
+                        builder.append("            <classifier>").append(a.getClassifier()).append("</classifier>").append(System.lineSeparator());
+                    }
+                    builder.append("            <version>").append(a.getVersion()).append("</version>").append(System.lineSeparator());
+                    builder.append("            <type>").append(a.getType()).append("</type>").append(System.lineSeparator());
+                    builder.append("          </artifact>").append(System.lineSeparator());
                 }
-                builder.append("        <version>").append(a.getVersion()).append("</version>").append(System.lineSeparator());
-                builder.append("        <type>").append(a.getType()).append("</type>").append(System.lineSeparator());
-                builder.append("      </artifact>").append(System.lineSeparator());
+                builder.append("        </module>").append(System.lineSeparator());
             }
-            builder.append("    </module>").append(System.lineSeparator());
+            builder.append("      </modules>").append(System.lineSeparator());
+            builder.append("    </feature-pack>").append(System.lineSeparator());
         }
-        builder.append("  </modules>").append(System.lineSeparator());
+        builder.append("  </feature-packs>").append(System.lineSeparator());
```

Comment:
Just a general comment here. Should we be using something like a `XMLStreamWriter` to ensure we get valid and escaped XML? If we want it nicely formatted we could use something like https://gist.github.com/jamezp/012fd06babecc39bb56e66d31d3896c6.